flink-runtime/src/main/java/org/apache/flink/runtime/state/proxy/ProxyKeyedStateBackend.java

```diff
@@ -0,0 +1,306 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state.proxy;
+
+import org.apache.flink.api.common.state.AggregatingStateDescriptor;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.api.common.state.ReducingStateDescriptor;
+import org.apache.flink.api.common.state.State;
+import org.apache.flink.api.common.state.StateDescriptor;
+import org.apache.flink.api.common.state.ValueStateDescriptor;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.state.AbstractKeyedStateBackend;
+import org.apache.flink.runtime.state.CheckpointStreamFactory;
+import org.apache.flink.runtime.state.KeyGroupedInternalPriorityQueue;
+import org.apache.flink.runtime.state.Keyed;
+import org.apache.flink.runtime.state.KeyedStateFunction;
+import org.apache.flink.runtime.state.KeyedStateHandle;
+import org.apache.flink.runtime.state.PriorityComparable;
+import org.apache.flink.runtime.state.SnapshotResult;
+import org.apache.flink.runtime.state.StateSnapshotTransformer;
+import org.apache.flink.runtime.state.heap.HeapKeyedStateBackend;
+import org.apache.flink.runtime.state.heap.HeapPriorityQueueElement;
+import org.apache.flink.runtime.state.internal.InternalKvState;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.IOUtils;
+
+import javax.annotation.Nonnull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.RunnableFuture;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** */
+public class ProxyKeyedStateBackend<K> extends AbstractKeyedStateBackend<K> {
+    // wrapped keyed state backend, either HeapKeyedStateBackend or RocksDBKeyedStateBackend
+    AbstractKeyedStateBackend<K> keyedStateBackend;
+
+    private static final Map<Class<? extends StateDescriptor>, StateFactory> STATE_FACTORIES =
+            Stream.of(
+                            Tuple2.of(
+                                    ValueStateDescriptor.class,
+                                    (StateFactory) ProxyValueState::create),
+                            Tuple2.of(
+                                    ListStateDescriptor.class,
```

Comment:
I do not have a preference in general, simpler is better. 

However, I am constructing the factory map this way to follow the same style/functionality existing keyedStateBackends use, since this is a wrapper for them. 

If you prefer to make them simpler, we probably should change them altogether. Having two different styles may lead to more confusion.

Reply:
I think different backends may have completely different implementations. I can imagine some backends **do** register state factories dynamically (and therefore need this complexity).
So I don't think it's necessary to for them to share the internals.

But if we **do** want this code to be shared then a better way would be to extract and re-use.

Reply:
1. I do not think it is overly complicated, it is using a mapper to store the create factory instead of the "if else statement"... To me, they are just different ways of implementation. Is the reason specifically you do not like this way because it is complicated?

2. I am a bit prefer this way because I want the wrapper to provide as least the same level of flexibility and extensiblity) as the wrapped one.

3. As I said, if you think it is not necessary, we can change to the non-dynamic way, but with all the wrapped version changed as well (as long as to keep them the same level of extensibility).

Reply:
1. Yes, because of unnecessary complexity (plus `instanceof` is more flexible than keys of `class` type).
2. Can you explain why is it more flexible and extensible than just `if-else`?
3. And why the same level of flexibility and extensibility is needed in wrapping and wrapped backends?

Reply:
> 1. Yes, because of unnecessary complexity (plus `instanceof` is more flexible than keys of `class` type).
> 2. Can you explain why is it more flexible and extensible than just `if-else`?. And why the same level of flexibility and extensibility is needed in wrapping and wrapped backends?
> 3. I'd prefer such a refactoring (if any) to be performed before adding a new backend

**Why more flexible and extensible:**
as you said, the mappings can be updated dynamically or programmatically, 
like `mapper.put (new type, new factory)`; 
and also easy for check `mapper.get(a type)` is existed or not. 

That's something if-else can not simply and cleanly do

**why the same level of flexibility and extensibility**
If the underlying can support dynamically add a new type, as what is now; I do want the wrapper to support "dynamically add a new type" as well.